[FSDP] Include duplicate parameters and modules when calling named_parameters and named_modules

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98912

The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

Differential Revision: [D44897935](https://our.internmc.facebook.com/intern/diff/D44897935/)

fegin committed:
[FSDP] Include duplicate parameters and modules when calling named_parameters and named_modules

The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

Differential Revision: [D44897935](https://our.internmc.facebook.com/intern/diff/D44897935/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Include duplicate parameters and modules when calling named_parameters and named_modules"


The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

Differential Revision: [D44897935](https://our.internmc.facebook.com/intern/diff/D44897935/)

[ghstack-poisoned]

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -1103,6 +1103,30 @@ def test_torch_save_load(self):
+        with FSDP.state_dict_type(model, StateDictType.FULL_STATE_DICT):
+            state_dict = model.state_dict()
+            for key, param in state_dict.items():
+                print(key)
```
nit: Do we want to keep the `print()` in the unit test? Maybe instead we check that the contents of the state dict are as expected, even by hard coding?

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -1103,6 +1103,30 @@ def test_torch_save_load(self):
                     self.assertEqual(v, state_dict[k])
 
+    @skip_if_lt_x_gpu(2)
+    def test_shared_module(self):
```
nit: Should we call this `test_shared_module_and_shared_param` or something? It looks like it has shared parameter as well.

awgu reviewed (approved):
LGTM! Maybe for the unit test, we might want to assert on the keys?

fegin committed:
Update on "[FSDP] Include duplicate parameters and modules when calling named_parameters and named_modules"


The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

Differential Revision: [D44897935](https://our.internmc.facebook.com/intern/diff/D44897935/)

[ghstack-poisoned]
